clusterversion: set `binaryMinSupportedVersion` to `v21_1`

ajwerner opened the pull request:
This probably should have been done when `Start21_2` was added.

Fixes #67277.

Release note: None

ajwerner committed:
clusterversion: set `binaryMinSupportedVersion` to `v21_1`

This probably should have been done when `Start21_2` was added.

Fixes #67277.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/67281)
<!-- Reviewable:end -->

rafiss commented:
Just stumbled into this; some process Qs:
1. do we need to prioritize this specific issue within one of the engineering teams?
2. do we need to add something to a process somewhere so some team (i guess release-eng?) does this for future releases?

knz commented:
does this need help?

ajwerner committed:
migration/migrationmanager: validate the cluster version before migration

Before this change, the code would run a migration, then validate that it
can bump the cluster version. This is a little bit wacky and can lead to
lots of jobs being created, jobs which may ultimately fail. This change
is good in that it both prevents jobs from being created and gives us an
error before we bump the cluster version to the fence version.

Release note (bug fix): Fixed a bug where migration jobs might run and
update a cluster version before the cluster was ready for the upgrade.
The bug could result in many extra failed migration jobs.

ajwerner commented:
This has been rebased and most of the failures removed because of https://github.com/cockroachdb/cockroach/pull/67822. 

I added another commit to fix a real problem. I am left somewhat confused: why are we running the migration before bumping to the fence version? @irfansharif am I missing something or is this migration call in the wrong place?

ajwerner committed:
migrations: fix TruncatedStateMigrationTest to inject the appropriate version

The test was only working because we allowed the minBinarySupportedVersion to be
quite old.

Release note: None

ajwerner commented:
>  @irfansharif am I missing something or is this migration call in the wrong place?

I think I've got it sorted. We do it to make sure the version gating for new nodes joining is sound

ajwerner commented:
I would like very much to backport the 2nd commit. Who needs to look at it and what additional testing does it need?

knz reviewed (commented):
1st, 3rd and 4th commit LGTM

Would like a review from @irfansharif on the 2nd commit too

> Reviewed 5 of 5 files at r4, 2 of 3 files at r5, 2 of 2 files at r7, 1 of 1 files at r8.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67281#-:-Mg6x2hXEgvf8C8fMQRO:b-rgbonp)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @ajwerner)

___
*[pkg/migration/migrationmanager/manager.go, line 295 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/67281#-Mg6uJvrBcmYEtpvSNGs:-Mg6uJvrBcmYEtpvSNGt:b-5kpo56) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb5f13a1cc2e0b01336fe99e7d52f40298ba0f2c/pkg/migration/migrationmanager/manager.go#L295)):*
> ```Go
> 		alreadyCompleted, err = migrationjob.CheckIfMigrationCompleted(ctx, txn, m.ie, version)
> 		if alreadyCompleted || err != nil {
> 			log.Infof(ctx, "already completed or error %v", err)
> ```

maybe split the cases and use `log.Errorf` where appropriate?


<!-- Sent from Reviewable.io -->

ajwerner reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/67281)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @knz)

___
*[pkg/migration/migrationmanager/manager.go, line 295 at r7](https://reviewable.io/reviews/cockroachdb/cockroach/67281#-Mg6uJvrBcmYEtpvSNGs:-MgB7HDy3oiOcAVs-yh7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/fb5f13a1cc2e0b01336fe99e7d52f40298ba0f2c/pkg/migration/migrationmanager/manager.go#L295)):*
<details><summary><i>Previously, knz (kena) wrote…</i></summary><blockquote>

maybe split the cases and use `log.Errorf` where appropriate?
</blockquote></details>

Done.


<!-- Sent from Reviewable.io -->

irfansharif reviewed (approved):
LGTM.

> why are we running the migration before bumping to the fence version? @irfansharif am I missing something or is this migration call in the wrong place?

> We do it to make sure the version gating for new nodes joining is sound

Yup. The other thing to note is that the release demarcation versions (`Start21_2`, `Start21_1`, etc.) don't have any migrations attached to them. The migration manager will then:
```go
runMigration(..., Start21_2)
bumpClusterVersion(..., fenceVersionFor(Start21_2))
runMigration(..., <first real version in 21.2>)
```

As I say this I realize we don't have panics/asserts making sure we don't accidentally attach a migration on one of these special `StartXY_Z` versions. We could document/assert in this PR if you feel it necessary.

irfansharif commented:
(Also sorry for the review lag, only now crawling out of my inbox after a hectic L2 rotation + a long weekend.)

ajwerner commented on pkg/migration/migrations/truncated_state_external_test.go:
```diff
@@ -39,21 +44,36 @@ func TestTruncatedStateMigration(t *testing.T) {
-				// We want to exercise manual control over the upgrade process.
-				DisableAutomaticVersionUpgrade: 1,
+			makeArgs := func() (args base.TestServerArgs) {
+				minVersion := minVersion
```
Detritus. I think at some point for some reason there were some pointers involved. Removed.

ajwerner commented:
TFTR!

bors r+
